modules/storages/app/services/storages/one_drive_managed_folder_sync_service.rb

```diff
@@ -0,0 +1,193 @@
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See COPYRIGHT and LICENSE files for more details.
+#++
+
+module Storages
+  class OneDriveManagedFolderSyncService
+    using Peripherals::ServiceResultRefinements
+
+    DISALLOWED_CHARS = /[\\<>+?:"|\/]/
+    OPG_PERMISSIONS = %i[read_files write_files create_files delete_files share_files].freeze
+
+    def self.call(storage)
+      new(storage).call
+    end
+
+    def initialize(storage)
+      @storage = storage
+    end
+
+    def call
+      return unless @storage.automatic_management_enabled?
+
+      rfm = remote_folders_map.on_failure { |failed_result| return failed_result }.result
+
+      ensure_folders_exist(rfm).on_success { hide_inactive_folders(rfm) }
+      apply_permission_to_folders
+    end
+
+    private
+
+    def apply_permission_to_folders
+      active_project_storages_scope.includes(:project).where.not(project_folder_id: nil).find_each do |project_storage|
+        permissions = { read: [], write: admin_client_tokens_scope.pluck(:origin_user_id) }
+        project_tokens(project_storage).each do |token|
+          add_user_to_permission_list(permissions, token, project_storage.project)
+        end
+
+        set_permissions(project_storage.project_folder_id, permissions)
+      end
+
+      ServiceResult.success
+    end
+
+    def ensure_folders_exist(folder_map)
+      active_project_storages_scope.includes(:project).find_each do |project_storage|
+        actual_path = project_folder_path(project_storage.project)
+        next create_folder(project_storage, actual_path) unless folder_map.key?(project_storage.project_folder_id)
+
+        if folder_map[project_storage.project_folder_id] != actual_path
+          rename_folder(project_storage.project_folder_id, actual_path)
+        end
+      end
+
+      ServiceResult.success(result: 'folders processed')
+    end
+
+    def hide_inactive_folders(folder_map)
+      project_folder_ids = active_project_storages_scope.pluck(:project_folder_id).compact
+      (folder_map.keys - project_folder_ids).each do |item_id|
+        Peripherals::Registry.resolve("commands.one_drive.set_permissions")
+                             .call(storage: @storage, path: item_id, permissions: { write: [], read: [] })
+                             .on_failure do |service_result|
+          format_and_log_error(service_result.errors, folder: path, context: 'hide_folder')
+        end
+      end
+    end
+
+    def add_user_to_permission_list(permissions, token, project)
+      opg_user_permissions = token.user.all_permissions_for(project)
+
+      if opg_user_permissions.member?(:write_files)
```

Comment:
@mereghost I haven't read the whole PR and just landed here. I wander what `opg` stands for. OpenProject greatness?

Reply:
`OpenProgect`? to be honest I don't remember by frame of mind at the time. `op_user_permissions` probably made me thing of `overpowered user permissions`, so I might have added the `g` there. 😛 

I suck at naming things. 🤣